com.unity.ml-agents/Runtime/Inference/TensorGenerator.cs

```diff
@@ -94,30 +94,39 @@ public interface IGenerator
         public void InitializeObservations(List<ISensor> sensors, ITensorAllocator allocator)
         {
             // Loop through the sensors on a representative agent.
-            // For vector observations, add the index to the (single) VectorObservationGenerator
-            // For visual observations, make a VisualObservationInputGenerator
+            // For vector observations, add the index to the (single) ObservationGenerator
+            // For visual observations, make a NonVectorObservationInputGenerator
             var visIndex = 0;
-            VectorObservationGenerator vecObsGen = null;
+            ObservationGenerator vecObsGen = null;
             for (var sensorIndex = 0; sensorIndex < sensors.Count; sensorIndex++)
             {
                 var sensor = sensors[sensorIndex];
                 var shape = sensor.GetObservationShape();
                 // TODO generalize - we currently only have vector or visual, but can't handle "2D" observations
-                var isVectorSensor = (shape.Length == 1);
-                if (isVectorSensor)
+                var rank = shape.Length;
+                switch (rank)
                 {
-                    if (vecObsGen == null)
-                    {
-                        vecObsGen = new VectorObservationGenerator(allocator);
-                    }
+                    case 1:
+                        if (vecObsGen == null)
+                        {
+                            vecObsGen = new ObservationGenerator(allocator);
+                        }
+                        vecObsGen.AddSensorIndex(sensorIndex);
+                        break;
+                    case 2:
+                        var gen = new ObservationGenerator(allocator);
+                        gen.AddSensorIndex(sensorIndex);
+                        m_Dict[TensorNames.ObservationPlaceholderPrefix + sensorIndex] = gen;
```

Comment:
Can you put something like "// NOTE observations with dimension 2 use the sensorIndex (which might have gaps) and visIndex (which doesn't have gaps)" ? I just want to make it clear that we're currently mixing two numbering schemes.